### ISSUE TITLE
Convolutions/poolings duplicate `tf.RaggedTensor` functionality?

### ISSUE THREAD

**Issue by jackd** [tensorflow/graphics#2]
First and foremost, thanks for the awesome package all.

I've been doing separate work on tf graph convolutions and it strikes me that a lot of the functionality in the convolutions/pooling is geared towards dealing with neighborhoods of irregular size. Isn't this essentially what `tf.RaggedTensor`s are intended to manage?

Spefically, couldn't functions with a `sizes` argument take a potentially ragged tensor instead? At first glance my feeling is a ragged tensor could always be constructed from the relevant padded `values`, `sizes` args, but in some cases construction of the padded `values` could be skipped entirely (though not using the current implementation). This can be significant when a small number of vertices have a very large number of nodes which can result in massive amounts of padding.

I'd have to spend a bit of time working through the details - and maybe there's some fuctionality I'm not considering - but I'd be happy to put together a pull request if it's something that would be considered.

**Comment by amakadia** [comment]
The first thing that happens within the convolution ops is that the optional batch dimensions are collapsed (which also removes any padding).

We chose to use a dense padded data tensor as opposed to a ragged tensor because it is more likely that the batched datasets would be dense/padded rather than ragged (I believe until very recently tf.data did not have support for RaggedTensors).

In any case it is always possible to collapse the data yourself before passing it in to the graph convolution ops.

**Comment by jackd** [comment]
Thanks @amakadia

> We chose to use a dense padded data tensor as opposed to a ragged tensor because it is more likely that the batched datasets would be dense/padded rather than ragged

I agree with this, but won't consecutive convolutions result in uneccessary padding and unpadding? Maybe some graph optimizer can do fancy things behind the scenes, but I feel it's asking a lot.

> until very recently tf.data did not have support for RaggedTensors

I was under the impression this was still the case (i.e. `tf.data` does not support `RaggedTensor`s), so I've written a fair bit of hacky code elsewhere to do this myself. I'm a big fan of deleting my own code though, so if you could point me in direction of this functionality I'd be terribly grateful.

> In any case it is always possible to collapse the data yourself before passing it in to the graph convolution ops.

Thanks, this is enough for my use cases. Having said that though, and looking more closely at the implementation I feel this just modifies my issue with the package to be "`utils.flatten_batch_to_2d` duplicates `RaggedTensor` functionality" - and, I suspect, inefficiently (based on my gut-feeling about `gather` and `scatter` efficiencies - I stand to be corrected). Could [this section](https://github.com/tensorflow/graphics/blob/master/tensorflow_graphics/geometry/convolution/utils.py#L275) not be more efficiently implemented as below?

```python
  if sizes is None:
      ...
    else:
      sizes_flat = tf.cast(tf.reshape(sizes, (-1,)), tf.int64)
      flat = tf.RaggedTensor.from_tensor(
        tf.reshape(data, tf.stack([-1, data_shape[-2], data_shape[-1]])),
        lengths=sizes_flat).flat_values

      def unflatten(flat, name=None):
        """Invert flatten_batch_to_2d."""
        with tf.compat.v1.name_scope(name, "utils_unflatten", [flat]):
          flat = tf.convert_to_tensor(value=flat)
          output_shape = tf.concat((data_shape[:-1], tf.shape(input=flat)[-1:]),
                                   axis=0)
          ragged = tf.RaggedTensor.from_row_lengths(flat, sizes_flat)
          padded = ragged.to_tensor()
          # the following padding is required at least for tests to pass
          # otherwise when sizes is all zeros you get an output the wrong size
          # the size discreprancy is limited to padded values, so whether this
          # is actually important is debatable
          padding = [[0, 0] for _ in range(3)]
          padding[1] = [0, output_shape[-2] - padded.shape[-2]]
          padded = tf.pad(padded, padding)
          return tf.reshape(padded, output_shape)
```

(All tests pass and I'm happy to submit a PR if the idea looks good in principle - just thought I'd float it here in case there were philosophical objections or if there are inefficiencies I'm not considering.)

**Comment by julienvalentin** [comment]
Hi @jackd ,

Thanks a lot for this; very cool!
Could you run a performance benchmark (memory + run-time) to assess how this compares to the current version?

Best.

**Comment by jackd** [comment]
Not sure on the best way to make this maintainable, but here's a dirty hack that will hopefully be enough for demonstration.

# TL;DR

### CPU/2.0

* the unpadded modified version is ~10x faster, padded version similar.
* Both modified versions require slightly more memory (~25%).

### GPU/1.13.1

* speed roughly the same
* Both modified values require more GPU memory (~40%) but negligible CPU memory (~113,000x less)

I guess that's less conclusive than I thought it would be... but I feel if there's a strong argument for the original implementation over the ragged implementation, then that same argument should be applied to change the relevant core tensorflow `ragged` implementation.

# Code

```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np
import tensorflow as tf

random = np.random.RandomState(123)

def flatten_batch_to_2d_original(data, sizes):
  """Original implementation."""
  data_shape = tf.shape(data)
  mask = tf.sequence_mask(sizes, data_shape[-2])
  mask_indices = tf.cast(tf.where(mask), tf.int32)
  flat = tf.gather_nd(params=data, indices=mask_indices)

  def unflatten(flat, name=None):
    """Invert flatten_batch_to_2d."""
    with tf.compat.v1.name_scope(name, "utils_unflatten", [flat]):
        flat = tf.convert_to_tensor(value=flat)
        output_shape = tf.concat((data_shape[:-1], tf.shape(input=flat)[-1:]),
                                    axis=0)
        return tf.scatter_nd(
            indices=mask_indices, updates=flat, shape=output_shape)

  return flat, unflatten


def flatten_batch_to_2d_ragged(data, sizes):
  """Implementation based on ragged tensors."""
  data_shape = tf.shape(data)
  sizes_flat = tf.cast(tf.reshape(sizes, (-1,)), tf.int64)
  flat = tf.RaggedTensor.from_tensor(
    tf.reshape(data, tf.stack([-1, data_shape[-2], data_shape[-1]])),
    lengths=sizes_flat).flat_values

  def unflatten(flat, name=None):
    """Invert flatten_batch_to_2d."""
    with tf.compat.v1.name_scope(name, "utils_unflatten", [flat]):
      flat = tf.convert_to_tensor(value=flat)
      output_shape = tf.concat((data_shape[:-2], [-1], tf.shape(input=flat)[-1:]),
                               axis=0)
      ragged = tf.RaggedTensor.from_row_lengths(flat, sizes_flat)
      padded = ragged.to_tensor()

      return tf.reshape(padded, output_shape)

  return flat, unflatten


def flatten_batch_to_2d_ragged_padded(data, sizes):
  """Same as `flatten_batch_to_2d_ragged` with additional padding."""
  data_shape = tf.shape(data)
  sizes_flat = tf.cast(tf.reshape(sizes, (-1,)), tf.int64)
  flat = tf.RaggedTensor.from_tensor(
    tf.reshape(data, tf.stack([-1, data_shape[-2], data_shape[-1]])),
    lengths=sizes_flat).flat_values

  def unflatten(flat, name=None):
    """Invert flatten_batch_to_2d."""
    with tf.compat.v1.name_scope(name, "utils_unflatten", [flat]):
      flat = tf.convert_to_tensor(value=flat)
      output_shape = tf.concat((data_shape[:-1], tf.shape(input=flat)[-1:]),
                               axis=0)
      ragged = tf.RaggedTensor.from_row_lengths(flat, sizes_flat)
      padded = ragged.to_tensor()
      padding = [[0, 0] for _ in range(3)]
      padding[1] = [0, data_shape[-2] - padded.shape[-2]]
      padded = tf.pad(padded, padding)

      return tf.reshape(padded, output_shape)

  return flat, unflatten


def run_benchmark(impl, data, sizes):
    name = impl.__name__
    with tf.Graph().as_default():
        data = tf.constant(data)
        sizes = tf.constant(sizes)

        flat, unflatten = impl(data, sizes)
        reshaped = unflatten(flat)

        with tf.compat.v1.Session() as sess:
            bm = tf.test.Benchmark()
            bm.run_op_benchmark(
                sess, flat, name='%s-flatten' % name)
            bm.run_op_benchmark(
                sess, reshaped, name='%s-flatten_and_unflatten' % name)

A = 1000
max_d1 = 1000
D2 = 10
data = random.uniform(size=(A, max_d1, D2)).astype(np.float32)
sizes = (random.uniform(size=(A,))*max_d1).astype(np.int64)
for impl in (
        flatten_batch_to_2d_original,
        flatten_batch_to_2d_ragged,
        flatten_batch_to_2d_ragged_padded):
    run_benchmark(impl, data, sizes)
```

# 2.0-alpha-CPU:

```
entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_original-flatten"
  iters: 10
  wall_time: 0.00257110595703125
  extras {
    key: "allocator_maximum_num_bytes_cpu"
    value {
      double_value: 20403720.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_original-flatten_and_unflatten"
  iters: 10
  wall_time: 0.06875348091125488
  extras {
    key: "allocator_maximum_num_bytes_cpu"
    value {
      double_value: 64484476.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_ragged-flatten"
  iters: 10
  wall_time: 0.0026935338973999023
  extras {
    key: "allocator_maximum_num_bytes_cpu"
    value {
      double_value: 20403720.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_ragged-flatten_and_unflatten"
  iters: 10
  wall_time: 0.00716710090637207
  extras {
    key: "allocator_maximum_num_bytes_cpu"
    value {
      double_value: 79760000.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_ragged_padded-flatten"
  iters: 10
  wall_time: 0.002659440040588379
  extras {
    key: "allocator_maximum_num_bytes_cpu"
    value {
      double_value: 20403720.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_ragged_padded-flatten_and_unflatten"
  iters: 10
  wall_time: 0.0475161075592041
  extras {
    key: "allocator_maximum_num_bytes_cpu"
    value {
      double_value: 80000000.0
    }
  }
}
```

# 1.13.1 GPU

```
entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_original-flatten"
  iters: 10
  wall_time: 0.00256609916687
  extras {
    key: "allocator_maximum_num_bytes_GPU_0_bfc"
    value {
      double_value: 60403720.0
    }
  }
  extras {
    key: "allocator_maximum_num_bytes_cuda_host_bfc"
    value {
      double_value: 4080744.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_original-flatten_and_unflatten"
  iters: 10
  wall_time: 0.0047914981842
  extras {
    key: "allocator_maximum_num_bytes_GPU_0_bfc"
    value {
      double_value: 60403720.0
    }
  }
  extras {
    key: "allocator_maximum_num_bytes_cuda_host_bfc"
    value {
      double_value: 4080756.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_ragged-flatten"
  iters: 10
  wall_time: 0.00215137004852
  extras {
    key: "allocator_maximum_num_bytes_GPU_0_bfc"
    value {
      double_value: 84080744.0
    }
  }
  extras {
    key: "allocator_maximum_num_bytes_cuda_host_bfc"
    value {
      double_value: 12.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_ragged-flatten_and_unflatten"
  iters: 10
  wall_time: 0.00447499752045
  extras {
    key: "allocator_maximum_num_bytes_GPU_0_bfc"
    value {
      double_value: 84080744.0
    }
  }
  extras {
    key: "allocator_maximum_num_bytes_cuda_host_bfc"
    value {
      double_value: 12.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_ragged_padded-flatten"
  iters: 10
  wall_time: 0.00215399265289
  extras {
    key: "allocator_maximum_num_bytes_GPU_0_bfc"
    value {
      double_value: 84080744.0
    }
  }
  extras {
    key: "allocator_maximum_num_bytes_cuda_host_bfc"
    value {
      double_value: 12.0
    }
  }
}

entry {
  name: "TensorFlowBenchmark.flatten_batch_to_2d_ragged_padded-flatten_and_unflatten"
  iters: 10
  wall_time: 0.0048885345459
  extras {
    key: "allocator_maximum_num_bytes_GPU_0_bfc"
    value {
      double_value: 84080744.0
    }
  }
  extras {
    key: "allocator_maximum_num_bytes_cuda_host_bfc"
    value {
      double_value: 36.0
    }
  }
}
```

**Comment by amakadia** [comment]
> 
> I agree with this, but won't consecutive convolutions result in uneccessary padding and unpadding? Maybe some graph optimizer can do fancy things behind the scenes, but I feel it's asking a lot.
> 
Hi @jackd,
This padding/unpadding (aka flatten/unflatten) inside the convolution op is only triggered when the input is batched. To compose multiple convolution ops one could flatten/unflatten before and after the conv ops:

```
data_flat, unflatten = utils.flatten_batch_to_2d(data_in, sizes)
neighbors_bd = utils.convert_to_block_diag_2d(neighbors, sizes_square)
data_flat2 = conv(data_flat, neighbors_bd, ...)
data_flat3 = conv(data_flat2, neighbors_bd, ...)
...
data_out = unflatten(data_flatN)
```
